### PR TITLE
Replace SubroutineCall when any of its arguments are updated in pass_array_by_data

### LINKED ISSUE #1486: ASR verify pass error for ASR from pass_array_by_data
```fortran
subroutine find_fit(expr)
interface
    real function expr(x)
        real, intent(in) :: x(:)
    end function expr
end interface
end subroutine find_fit

program example_primes
implicit none
call find_fit(expression)

contains
real function expression(x)
    real, intent(in) :: x(:)
end function expression

end program example_primes
```

> LFortran

```console
$ lfortran a.f90 
ASR verify pass error: ASR verify: Var::m_v `expression` cannot point outside of its symbol table
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/home/pranavchiku/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main_impl()
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main()
  File "/home/pranavchiku/lfortran/src/bin/lfortran.cpp", line 1946, in ??
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/home/pranavchiku/lfortran/src/bin/lfortran.cpp", line 867, in ??
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/home/pranavchiku/lfortran/src/lfortran/fortran_evaluator.cpp", line 342, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 8035, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    pass_manager.apply_passes(al, &asr, pass_options, diagnostics);
  File "/home/pranavchiku/lfortran/src/libasr/pass/pass_manager.h", line 253, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    _apply_passes(al, asr, _passes, pass_options, diagnostics);
LCompilersException: Verify failed
```

### LINKED ISSUE #1488: pass_array_by_data does not checks nested functions if pass array by data is not possible in parent function
```fortran
program example_primes
implicit none

contains
subroutine find_fit(expr)
interface
    real function expr(x)
        real, intent(in) :: x(:)
    end function expr
end interface
end subroutine find_fit

end program example_primes
```

<details>
<summary> ASR Generated after pass_array_by_data </summary>
<p>

```clojure
(TranslationUnit
    (SymbolTable
        1
        {
            example_primes:
                (Program
                    (SymbolTable
                        2
                        {
                            find_fit:
                                (Function
                                    (SymbolTable
                                        3
                                        {
                                            expr:
                                                (Function
                                                    (SymbolTable
                                                        4
                                                        {
                                                            expr:
                                                                (Variable
                                                                    4
                                                                    expr
                                                                    []
                                                                    ReturnVar
                                                                    ()
                                                                    ()
                                                                    Default
                                                                    (Real 4 [])
                                                                    Source
                                                                    Public
                                                                    Required
                                                                    .false.
                                                                ),
                                                            x:
                                                                (Variable
                                                                    4
                                                                    x
                                                                    []
                                                                    In
                                                                    ()
                                                                    ()
                                                                    Default
                                                                    (Real 4 [(()
                                                                    ())])
                                                                    Source
                                                                    Public
                                                                    Required
                                                                    .false.
                                                                )
                                                        })
                                                    expr
                                                    (FunctionType
                                                        [(Real 4 [(()
                                                        ())])]
                                                        (Real 4 [])
                                                        Source
                                                        Interface
                                                        ()
                                                        .false.
                                                        .false.
                                                        .false.
                                                        .false.
                                                        .false.
                                                        []
                                                        []
                                                        .false.
                                                    )
                                                    []
                                                    [(Var 4 x)]
                                                    []
                                                    (Var 4 expr)
                                                    Public
                                                    .false.
                                                    .false.
                                                )
                                        })
                                    find_fit
                                    (FunctionType
                                        [(FunctionType
                                            [(Real 4 [(()
                                            ())])]
                                            (Real 4 [])
                                            Source
                                            Interface
                                            ()
                                            .false.
                                            .false.
                                            .false.
                                            .false.
                                            .false.
                                            []
                                            []
                                            .false.
                                        )]
                                        ()
                                        Source
                                        Implementation
                                        ()
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        []
                                        .false.
                                    )
                                    []
                                    [(Var 3 expr)]
                                    []
                                    ()
                                    Public
                                    .false.
                                    .false.
                                )
                        })
                    example_primes
                    []
                    []
                )
        })
    []
)
```

</p>
</details>

<details>
<summary> Correct ASR </summary>
<p>

```console
(TranslationUnit
    (SymbolTable
        1
        {
            example_primes:
                (Program
                    (SymbolTable
                        2
                        {
                            find_fit:
                                (Function
                                    (SymbolTable
                                        3
                                        {
                                            expr_x:
                                                (Function
                                                    (SymbolTable
                                                        5
                                                        {
                                                            __1x:
                                                                (Variable
                                                                    5
                                                                    __1x
                                                                    []
                                                                    In
                                                                    ()
                                                                    ()
                                                                    Default
                                                                    (Integer 4 [])
                                                                    Source
                                                                    Public
                                                                    Required
                                                                    .false.
                                                                ),
                                                            __2x:
                                                                (Variable
                                                                    5
                                                                    __2x
                                                                    []
                                                                    In
                                                                    ()
                                                                    ()
                                                                    Default
                                                                    (Integer 4 [])
                                                                    Source
                                                                    Public
                                                                    Required
                                                                    .false.
                                                                ),
                                                            expr:
                                                                (Variable
                                                                    5
                                                                    expr
                                                                    []
                                                                    ReturnVar
                                                                    ()
                                                                    ()
                                                                    Default
                                                                    (Real 4 [])
                                                                    Source
                                                                    Public
                                                                    Required
                                                                    .false.
                                                                ),
                                                            x:
                                                                (Variable
                                                                    5
                                                                    x
                                                                    [__1x
                                                                    __2x]
                                                                    In
                                                                    ()
                                                                    ()
                                                                    Default
                                                                    (Real 4 [((Var 5 __1x)
                                                                    (Var 5 __2x))])
                                                                    Source
                                                                    Public
                                                                    Required
                                                                    .false.
                                                                )
                                                        })
                                                    expr_x
                                                    (FunctionType
                                                        [(Real 4 [(()
                                                        ())])]
                                                        (Real 4 [])
                                                        Source
                                                        Interface
                                                        ""
                                                        .false.
                                                        .false.
                                                        .false.
                                                        .false.
                                                        .false.
                                                        []
                                                        []
                                                        .false.
                                                    )
                                                    []
                                                    [(Var 5 x)
                                                    (Var 5 __1x)
                                                    (Var 5 __2x)]
                                                    []
                                                    (Var 5 expr)
                                                    Public
                                                    .false.
                                                    .false.
                                                )
                                        })
                                    find_fit
                                    (FunctionType
                                        [(FunctionType
                                            [(Real 4 [(()
                                            ())])]
                                            (Real 4 [])
                                            Source
                                            Interface
                                            ()
                                            .false.
                                            .false.
                                            .false.
                                            .false.
                                            .false.
                                            []
                                            []
                                            .false.
                                        )]
                                        ()
                                        Source
                                        Implementation
                                        ()
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        []
                                        .false.
                                    )
                                    []
                                    [(Var 3 expr)]
                                    []
                                    ()
                                    Public
                                    .false.
                                    .false.
                                )
                        })
                    example_primes
                    []
                    []
                )
        })
    []
)
```
</p>
</details>

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1486. Fixes #1488.

### CODE REVIEW

File: integration_tests/pass_array_by_data_03.f90 [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        real, intent(in) :: x(3)
+    end function expr
+end interface
+
+x(1) = 1.0
+x(2) = 2.0
+x(3) = 3.0
+
+y = expr(x)
+if (abs(y - 6.0) > 1e-16) error stop
```
Comment (czgdp1807) [inline integration_tests/pass_array_by_data_03.f90:26]: Good.

File: src/libasr/pass/pass_array_by_data.cpp [hunk 2]
```diff
@@ -250,6 +250,34 @@ class PassArrayByDataProcedureVisitor : public PassUtils::PassVisitor<PassArrayB
...
+                                        ASR::down_cast<ASR::Var_t>(xx.m_args[i])->m_v = sym; 
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
```
Comment (czgdp1807) [inline src/libasr/pass/pass_array_by_data.cpp:280]: Exactly the same code as `visit_Function`?

Comment (Pranavchiku) [reply src/libasr/pass/pass_array_by_data.cpp:280]: Slightly different, in this, I update the argument of parent function to point to updated procedure.

File: integration_tests/pass_array_by_data_03.f90 (hunk 1 above)
Comment (certik) [inline integration_tests/pass_array_by_data_03.f90:26]: Single precision:
```suggestion
if (abs(y - 6.0) > 1e-7) error stop
```

**Review by certik — APPROVED** [review]
I think this is fine.

**Pull request merged**